lib/jsonpatch/utils.ex

```diff
@@ -0,0 +1,349 @@
+        end
+
+      {:custom, custom_fn} ->
+        case custom_fn.(fragment, path, target, opts) do
+          {:ok, _} = ok -> ok
+          :error -> {:error, {:invalid_path, path ++ [fragment]}}
+        end
+    end
+  end
+
+  @doc """
+  Uses a JSON patch path to get the last map/list that this path references.
+
+  ## Examples
+
+      iex> path = "/a/b/c/d"
+      iex> target = %{"a" => %{"b" => %{"c" => %{"d" => 1}}}}
+      iex> Jsonpatch.Utils.get_destination(target, path)
+      {:ok, {%{"d" => 1}, "d"}}
+
+      iex> # Invalid path
+      iex> path = "/a/e/c/d"
+      iex> target = %{"a" => %{"b" => %{"c" => %{"d" => 1}}}}
+      iex> Jsonpatch.Utils.get_destination(target, path)
+      {:error, {:invalid_path, ["a", "e"]}}
+
+      iex> path = "/a/b/1/d"
+      iex> target = %{"a" => %{"b" => [true, %{"d" => 1}]}}
+      iex> Jsonpatch.Utils.get_destination(target, path)
+      {:ok, {%{"d" => 1}, "d"}}
+
+      iex> path = "/a/b/1"
+      iex> target = %{"a" => %{"b" => [true, %{"d" => 1}]}}
+      iex> Jsonpatch.Utils.get_destination(target, path)
+      {:ok, {[true, %{"d" => 1}], 1}}
+
+      iex> # Invalid path
+      iex> path = "/a/b/42/d"
+      iex> target = %{"a" => %{"b" => [true, %{"d" => 1}]}}
+      iex> Jsonpatch.Utils.get_destination(target, path)
+      {:error, {:invalid_path, ["a", "b", "42"]}}
+  """
+
+  @spec get_destination(
+          target :: Types.json_container(),
+          path :: String.t(),
+          Types.opts()
+        ) ::
+          {:ok, {Types.json_container(), last_fragment :: Types.casted_fragment()}}
+          | Types.error()
+  def get_destination(target, path, opts \\ []) do
+    with {:ok, fragments} <- split_path(path) do
+      find_destination(target, [], fragments, opts)
+    end
+  end
+
+  @doc """
+  Updatest a map/list reference by a given JSON patch path with the new destination.
+
+  ## Examples
+
+      iex> path = "/a/b/c/d"
+      iex> target = %{"a" => %{"b" => %{"c" => %{"xx" => 0, "d" => 1}}}}
+      iex> Jsonpatch.Utils.update_destination(target, %{"e" => 1}, path)
+      {:ok, %{"a" => %{"b" => %{"c" => %{"e" => 1}}}}}
+
+      iex> path = "/a/b/1"
+      iex> target = %{"a" => %{"b" => [0, 1, 2]}}
+      iex> Jsonpatch.Utils.update_destination(target, 9999, path)
+      {:ok, %{"a" => %{"b" => 9999}}}
+  """
+  @spec update_destination(
+          target :: Types.json_container(),
+          value :: term(),
+          String.t(),
+          Types.opts()
+        ) ::
+          {:ok, Types.json_container()} | Types.error()
+  def update_destination(target, value, path, opts \\ []) do
+    with {:ok, fragments} <- split_path(path) do
+      do_update_destination(target, value, [], fragments, opts)
+    end
+  end
+
+  @doc """
+  Unescape `~1` to  `/` and `~0` to `~`.
+  """
+  @spec unescape(fragment :: String.t() | integer()) :: String.t()
+  def unescape(fragment) when is_binary(fragment) do
+    fragment
```

Comment:
The `unescape` should contain a check if the replacing is necessary. See https://github.com/corka149/jsonpatch/issues/11

Reply:
```elixir
:timer.tc(fn ->
  Enum.each(1..1_000_000, fn _ -> String.replace("uhuhuhuhuhuhuhuhuhuhuhuhuhuhhuhuhu", "~0", "/") end)
end)
{1445115, :ok}

:timer.tc(fn ->
  Enum.each(1..1_000_000, fn _ -> String.contains?("uhuhuhuhuhuhuhuhuhuhuhuhuhuhhuhuhu", "~0") end)
end)
{1114837, :ok}
```

So `String.replace` is about 30% slower that  `String.contains?` (in this toy example, on my machine).
But using `String.contains?` imply a second scan with `String.replace`, in the worst case (ie if the string should be escaped). In this case we will pay 1114837 + 1445115 -> 70% slower than single `String.replace`.

So this kind of optimisation depend on the probability distribution of the input string being in the set of string that need an escape or not.

IMHO favouring one and saving 30% or the other and saving 70% is quite arbitrary.

BTW I'm OK to revert, in our use case the impact of the library processing time is very small whatever input JSON input we patch.

PS: the change was not done with this reasoning it was a just a spontaneous code simplification at the time of review with my colleagues ;).